units/log_output_tests.cc

```diff
@@ -0,0 +1,167 @@
+    // time, it's opened in the append mode.
+    out_ = fopen(filename, "a");
+    if (!out_) {
+      SPL_LOG(ERROR) << "Failed to open " << filename
+                     << ". Using stderr as the alternative output.";
+      out_ = stderr;
+    }
+  }
+
+  ~FileOutput() {
+    if (out_ && out_ != stderr) {
+      fclose(out_);
+    }
+  }
+
+  void Flush() override {
+    assert(out_);
+    fflush(out_);
+  }
+
+  void LogLine(const std::string_view &line) override {
+    assert(out_);
+    assert(!line.find('\n'));
+    fprintf(out_, "%s\n", line.data());
+    Flush();
+  }
+
+ private:
+  FILE *out_ = nullptr;
+};
+
+// This class is used for testing. It writes the data to a string
+// instead of a file. The data can be read using the `GetLog()` method.
+class StringOutput : public LogOutput {
+ public:
+  StringOutput(){};
+
+  void Flush() override {}
+
+  void LogLine(const std::string_view &line) override {
+    assert(!line.find('\n'));
+    out_.push_back(line);
+  }
+
+  const std::vector<std::string_view> &GetLog() { return out_; }
+
+ private:
+  std::vector<std::string_view> out_;
+};
+
+// Used by the tests prior to logging to make sure the underlying file is empty.
+void CreateEmptyFile(std::string &filename) { std::ofstream(filename.c_str()); }
+
+TEST(LogOutput, SingleThreadedLog) {
+  std::string file_path = TempDir() + "/single_thread.log";
+  CreateEmptyFile(file_path);
+
+  const std::string line = "name:event_name,timestamp:1234";
+  FileOutput file_out(file_path.c_str());
+  file_out.LogLine(line);
+
+  std::ifstream test_file;
+  test_file.open(file_path);
+  EXPECT_TRUE(test_file.is_open());
+  std::string stored_line;
+  getline(test_file, stored_line);
+  EXPECT_EQ(stored_line, line);
+  test_file.close();
+
+  StringOutput string_out;
+  string_out.LogLine(line);
+  EXPECT_THAT(string_out.GetLog(), ElementsAre(line));
+}
+
+// This function is executed by multiple threads. It logs the input to the file
+// as many times as `kLogsPerThread`.
+void LogToFileOutput(FileOutput *file_out, const std::string &line) {
+  for (size_t i = 0; i < kLogsPerThread; ++i) file_out->LogLine(line);
+}
+
+TEST(LogOutput, MultiThreadedFileLog) {
+  std::string file_path = TempDir() + "/multi_thread.log";
+  CreateEmptyFile(file_path);
+
+  FileOutput file_out(file_path.c_str());
+  const std::string line = "name:event_name,timestamp:1234";
+  std::array<std::thread, kThreadCount> threads;
+  for (std::thread &thread : threads)
+    thread = std::thread(LogToFileOutput, &file_out, line);
+
+  for (std::thread &thread : threads) thread.join();
+
+  std::ifstream test_file;
+  test_file.open(file_path);
+  EXPECT_TRUE(test_file.is_open());
+  std::string stored_line;
+  for (size_t i = 0; i < kThreadCount * kLogsPerThread; ++i) {
+    getline(test_file, stored_line);
+    EXPECT_EQ(stored_line, line);
+  }
+  getline(test_file, stored_line);
+  EXPECT_EQ(stored_line, "");
+  test_file.close();
```

Comment:
no need to close manually, the destructor will handle this: https://en.cppreference.com/w/cpp/io/basic_fstream#:~:text=(destructor),virtual%5D%20(implicitly%20declared)